### PR TITLE
[SPARK-37827][SQL] Put the some built-in table properties into V1Table.propertie to adapt to V2 command

### PR DESCRIPTION
**Pull request by Peng-Lei**
### What changes were proposed in this pull request?
Add V2 table build-in properties into V1Table.properties to adapt to V2 command.

### Why are the changes needed?
[#discuss](https://github.com/apache/spark/pull/34773#issuecomment-1005857748)

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add ut testcase and existed ut testcase.

### CODE REVIEW

**Comment by Peng-Lei** [comment]
@cloud-fan @imback82 @huaxingao Could you take a look? Thank you very much.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 1]
```diff
@@ -384,9 +384,12 @@ case class CatalogTable(
   def toLinkedHashMap: mutable.LinkedHashMap[String, String] = {
     val map = new mutable.LinkedHashMap[String, String]()
     val tableProperties = properties
-      .filterKeys(!_.startsWith(VIEW_PREFIX))
+      .filter(!_._1.startsWith(VIEW_PREFIX))
+      .filter(kv => !CatalogV2Util.TABLE_RESERVED_PROPERTIES.contains(kv._1))
+      .filter(!_._1.startsWith(TableCatalog.OPTION_PREFIX))
+      .filter(_._1 != TableCatalog.PROP_EXTERNAL)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:390]: hmm, isn't `PROP_EXTERNAL` included in `TABLE_RESERVED_PROPERTIES`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 2]
```diff
@@ -384,9 +384,12 @@ case class CatalogTable(
...
       .toSeq.sortBy(_._1)
-      .map(p => p._1 + "=" + p._2).mkString("[", ", ", "]")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:389]: why this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala [hunk 3]
```diff
@@ -55,7 +56,8 @@ private[sql] case class V1Table(v1Table: CatalogTable) extends Table {
     }
   }
 
-  override lazy val properties: util.Map[String, String] = v1Table.properties.asJava
+  override lazy val properties: util.Map[String, String] =
+    V1Table.filterTableProperties(v1Table).asJava
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala:60]: I think we should generate these v2 table properties here, instead of when creating the table. It's a breaking change to generate v2 properties when creating tables, as it changes what to store in the metastore, which may have unknown consequences.

Comment (Peng-Lei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala:60]: Thank you very much. I try do it. In fact, officially because store these v2 table properties into metastore when create table, I spent a lot of time looking for failed use cases and trying to fix. It is indeed a breaking change to generate v2 properties when creating tables and store them. Thank you for your reminder very much. @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 1 above)
Comment (Peng-Lei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:390]: yes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 2 above)
Comment (Peng-Lei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:389]: original code
```scala
val tableProperties = properties
      .filterKeys(!_.startsWith(VIEW_PREFIX))
      .toSeq.sortBy(_._1)
      .map(p => p._1 + "=" + p._2).mkString("[", ", ", "]")
...
if (properties.nonEmpty) map.put("Table Properties", tableProperties) // Here I think it is more reasonable to judge the empty of tablePropertie, but it is string "[]", not empty.
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:389]: makes sense

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 4]
```diff
@@ -414,7 +414,8 @@ case class CatalogTable(
       }
     }
 
-    if (properties.nonEmpty) map.put("Table Properties", tableProperties)
+    if (tableProperties.nonEmpty) map.put("Table Properties",
+      tableProperties.mkString("[", ", ", "]"))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:418]: nit:
```
if (...) {
  ...
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala [hunk 5]
```diff
@@ -82,6 +84,21 @@ private[sql] case class V1Table(v1Table: CatalogTable) extends Table {
...
+    val external = v1Table.tableType == CatalogTableType.EXTERNAL
+
+    v1Table.properties ++
+      v1Table.storage.properties.map { case (key, value) =>
+        TableCatalog.OPTION_PREFIX + key -> value } ++
+      v1Table.provider.map(TableCatalog.PROP_PROVIDER -> _) ++
+      v1Table.comment.map(TableCatalog.PROP_COMMENT -> _) ++
+      v1Table.storage.locationUri.map(TableCatalog.PROP_LOCATION -> _.toString) ++
+      (if (external) Some(TableCatalog.PROP_EXTERNAL -> "true") else None) ++
+      Map(TableCatalog.PROP_OWNER -> Utils.getCurrentUserName())
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala:98]:
```suggestion
      Map(TableCatalog.PROP_OWNER -> v1Table.owner)
```

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 6]
```diff
@@ -2899,6 +2899,39 @@ class DataSourceV2SQLSuite
...
+           |LOCATION '/tmp'
+        """.stripMargin)
+
+      val table = spark.sessionState.catalogManager.v2SessionCatalog.asTableCatalog
+        .loadTable(Identifier.of(Array("default"), t))
+      val properties = table.properties
+      assert(properties.get(TableCatalog.PROP_PROVIDER) == "parquet")
+      assert(properties.get(TableCatalog.PROP_COMMENT) == "This is a comment")
+      assert(properties.get(TableCatalog.PROP_LOCATION) == "file:/tmp")
+      assert(properties.get(TableCatalog.PROP_OWNER) == Utils.getCurrentUserName())
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2927]: This is a bit unreliable. I'd test `assert(properties.containsKey(TableCatalog.PROP_OWNER))`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala [hunk 7]
```diff
@@ -118,11 +118,12 @@ class V2SessionCatalogTableSuite extends V2SessionCatalogBaseSuite {
     assert(!catalog.tableExists(testIdent))
 
     val table = catalog.createTable(testIdent, schema, Array.empty, emptyProps)
+    val loaded = catalog.loadTable(testIdent)
 
     val parsed = CatalystSqlParser.parseMultipartIdentifier(table.name)
     assert(parsed == Seq("db", "test_table"))
     assert(table.schema == schema)
-    assert(table.properties.asScala == Map())
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala:125]: can we keep the empty properties check?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala [hunk 8]
```diff
@@ -253,16 +255,14 @@ class V2SessionCatalogTableSuite extends V2SessionCatalogBaseSuite {
     val catalog = newCatalog()
 
     val table = catalog.createTable(testIdent, schema, Array.empty, emptyProps)
+    val loaded = catalog.loadTable(testIdent)
 
-    assert(table.properties.asScala == Map())
+    assert(table.properties == loaded.properties)
 
     val updated = catalog.alterTable(testIdent, TableChange.setProperty("prop-1", "1"))
-    assert(updated.properties.asScala == Map("prop-1" -> "1"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala:260]: ditto, can we keep this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 4 above)
Comment (Peng-Lei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala (hunk 5 above)
Comment (Peng-Lei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 6 above)
Comment (Peng-Lei) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala (hunk 7 above)
Comment (Peng-Lei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala:125]: done.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala (hunk 8 above)
Comment (Peng-Lei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala:260]: done

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
